## DataQuest Guided project
---

# GooglePlay and AppStore applications analisys. 

> this project is my first step in data analisys

**Remarks**

As it stated in title this is guided project, meaning to say that I follow advices rather than do my own research. Anyways there are some of my own-designed methods are implemented which I hope won't be judged too strictly.

**Legend**

I am working as data analist in GameDev company. The company develops free Apps and earn muney from adds. So, basicaly rewenue of company fully depends on amount of users. Company targeting on english-speaking users.

**Goal**

To analyze App Store and Google Play data to help developers understand what type of apps are likely to attract more users and consequently increase companies revenue.

**Tools**
1. Dataset - slice of data representing 10000 Android apps from Google Play;
2. Dataset - slice of data representing 7000 iOS apps from the App Store;

# Step 1. Data import and initial assesment
 * code below designed to open dataset in form allowing us to make firss assesment and arrange a workflow;

In [27]:
from csv import reader

open_AppStore = open('AppleStore.csv', encoding='utf-8', newline='')
read_AppStore = reader(open_AppStore)
AppStore_Dataset = list(read_AppStore)
AppStore_header = AppStore_Dataset[0]
AppStore_data = AppStore_Dataset[1:]

open_GooglePlay = open('googleplaystore.csv', encoding='utf-8', newline='')
read_GooglePlay = reader(open_GooglePlay)
GooglePlay_Dataset = list(read_GooglePlay)
GooglePlay_header = GooglePlay_Dataset[0]
GooglePlay_data = GooglePlay_Dataset[1:]

def explore_data(dataset, start, end, rows_and_columns = False):
    slice_dataset = dataset[start:end]
    for row in slice_dataset:
        print(row)
        print('\n')
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

 * import `AppStore,csv` and assess which data we might need for our research

In [28]:
print(AppStore_header)
explore_data(AppStore_data, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


* `AppStore.csv` is obtained from [keegle](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps). Check for column data explanation. For our research we can use `id`, `track_name`, `currency`, `price`, `rating_count_tot`,`user_rating`, `cont_rating` and `prime_genre`  columns.
---
* Performing the same steps for `googleplaystore.csv`:

In [29]:
print(GooglePlay_header)
explore_data(GooglePlay_data, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


* for our research we can use `app`, `category`, `rating`, `Installs`, `Type`, `Price`, `Content Rating` and `Genres` column data. 

# Step 2. Data checking and cleaning

1. Lets create function which will check if data is complete

* idea behind that function is that it loops trough the list and search for incomplete rows and missing data;
* function returns cleared dataset. This is first stage of filtering;

*might look a bit explicit*


In [30]:
def data_check(dataset, header):
    counter = 0
    i = -1
    for row in dataset:
        i += 1
        if len(row) != len(header):
            print('Row is incomplete for ' + row[0] + ' and was deleted')
            print(row)
            dataset.pop(i)
            counter += 1
        for element in row:
            if element == '' or None:
                print(header[row.index('')] + ' is missing for ' + row[0])
                print(row)
                counter += 1
    if counter == 0:
        print('Dataset complete!')
    return dataset

Perform data check for both datasets:

In [31]:
AppStore_data = data_check(AppStore_data, AppStore_header) #Appstore data

Dataset complete!


In [32]:
GooglePlay_data = data_check(GooglePlay_data, GooglePlay_header) #GooglePlay data

Current Ver is missing for Market Update Helper
['Market Update Helper', 'LIBRARIES_AND_DEMO', '4.1', '20145', '11k', '1,000,000+', 'Free', '0', 'Everyone', 'Libraries & Demo', 'February 12, 2013', '', '1.5 and up']
Row is incomplete for Life Made WI-Fi Touchscreen Photo Frame and was deleted
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
Content Rating is missing for Life Made WI-Fi Touchscreen Photo Frame
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


To test if dataset was actually modified lets check if length of dataset changed. Remember that before GooglePlay Store dataset contained 10841 rows of data.

In [33]:
print(len(GooglePlay_data))

10840


AppStore data was complete, when GooglePlay data was modefied. **Current version** missing is not critical for us, as this is data is not of our concern.

2. Next step of data cleaning is to check for duplicate data. 

I will start from checking our intial dataset for duplicate entries and will decide aftewards the secuence of steps. Might be wise to clean data first and then sort out duplicate entries. Might be vice versa. Probably it is matter of tastes. Let's start, as we say  `menshe slov, bolshe dela`

In [34]:
def duplicate_check(dataset):
    unique_names = []
    duplicates = []
    duplicate_ones = {}
    for app in dataset:
        name = app[0]
        if name not in unique_names:
            unique_names.append(name)
        else:
            duplicates.append(app)
            duplicate_ones[name] = 'badass'
    print('number of unique apps in dataset: ' + str(len(unique_names)))
    print('number of duplicate apps in dataset: ' + str(len(duplicates)))
    if len(duplicates) > 0:
        print('\n')
        print(len(duplicate_ones)) 

In [35]:
duplicate_check(AppStore_data)

number of unique apps in dataset: 7197
number of duplicate apps in dataset: 0


In [36]:
duplicate_check(GooglePlay_data)

number of unique apps in dataset: 9659
number of duplicate apps in dataset: 1181


798


This pretty exagerrated check shows us that AppStore data is totaly fine, when GooglePlay store data has 1181 duplicate entries. As I am using VScode and I am allowed to use all tools in my hand. So, I checked dataset for Google and noticed that some cases of double entries are caused by `free` and `paid` status of app. That means that probably before fighting with windmeals - better to clean data and sort out duplicates later.

Lets create new dataset of free apps:
* For AppStore data: if column[4] == 0.0 is true, than we leave app in dataset, otherwise delete;
* For GooglePlay data: if column [6] == 'free' is true, than we leave app in dataset, otherwise delete;

In [37]:
ios_free = []
for row in AppStore_data:
    if row[4] == '0.0':
        ios_free.append(row)
print (len(ios_free))

4056


In [38]:
android_free = []
for row in GooglePlay_data:
    if row[6] == 'Free':
        android_free.append(row)
print(len(android_free))

10039


As we are targeting in english speaking audience who prefer free apps - we will clean data taking that into account. To find out non-english apps is a bit challenging. Here I am out of knowledge, and had to seek for advice. Advice states:

> One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

> All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

> Some app's names consist not only from characters but also emojis and prefixes, we need to find way to leave those apps in for our analysis. One way to sort this issue is to make allowance for those characters outside abovementioned range - 127.

Having abovementiond in mind:

In [39]:
ios_free_eng = []
for row in ios_free:
    non_ASCII = 0
    for character in row[1]:
        if ord(character) > 127:
            non_ASCII += 1
    if non_ASCII < 4:
        ios_free_eng.append(row)
print(len(ios_free_eng))
         
    

3222


In [40]:
android_free_eng = []
for row in android_free:
    non_ASCII = 0
    for character in row[0]:
        if ord(character) > 127:
            non_ASCII += 1
    if non_ASCII < 4:
        android_free_eng.append(row)
print(len(android_free_eng))
         
    

9998


At this point I am ready to check again how many duplicate entries are in our new dataset for android apps. Before there were 1181 duplicate entries.

In [41]:
duplicate_check(android_free_eng)

number of unique apps in dataset: 8863
number of duplicate apps in dataset: 1135


756


Unfortunately there are still a lot. We got 1135 duplicate entries with 756 unique names of those. To make a final cleaning we will create a dictionary and invent criterion to sort which of duplicates we will keep for analisys. As I am using VScode IDE, I will investigate and report to you findings in next cell. 

Hey there. After close look on data in duplicate entries I found out that the only difference is number of rewievs. I don't know what is the algorythm of that segregation. Might be ерфе it devideы reviews by geography, might be that it does by time of the day. The only thing which is clear is that the more rewiews the more valuable and concious data inhereted in Rating. I have my doubts about this approach, cause we targeting on englishspeaking audience. For example Instagram can be used in local language, and if duplicate app data segrigated by geography, than we take into account irrelevant data, cause I can assume, that there are more asian people using this app, as there are simply more people who lives in Asia. But this can't be checked in scope of this research and I will keep it simple. Let's write code to create new dataset which will be last step in datacleaning - will remove duplicates from android apps data.

In [42]:
revievs_max = {}
for app in android_free_eng:
    name = app[0]
    n_reviews = float(app[3])
    if name not in revievs_max:
        revievs_max[name] = n_reviews
    elif revievs_max[name] > n_reviews:
        revievs_max[name] = n_reviews
print(len(revievs_max))

8863


8863 is number of unique apps in dictionary. Number is equal to one we obtained before, means that filtration went well.

At this poin we ready to create final dataset for GooglePlay. I intend to use created dictionary with max_reviev number to fill in new list lf lists by rerieving data based on apps name and review number.

In [43]:
android_final = []
already_added = []
for app in android_free_eng:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == revievs_max[name] and name not in already_added:
        android_final.append(app)
        already_added.append(name)
print(len(android_final))

8863


Congratulations! We finaly end up having clean datasets `android_final` and `ios_free_eng`

# Step 3. Research data analysis

At this point we ready to start data analysis, but before some explanations. Our company's workflow can be described as follows:

1. Build a minimal Android version of the app, and add it to Google Play;
2. If the app has a good response from users, we develop it further;
3. If the app is profitable after six month, we build an IOS version of the app  and add it to the App Store.

Before building up a minimall Android version we have to find out which app genre is most popular. For that goal we will analise GooglePlay dataset column 9 `genres` and 1 `category`; for AppStore dataset column 11 `prime genres`.

For that purpose we will write function which will return app genres and percentage of occurance in descending order.

In [54]:
def freq_table(dataset, index):
    frequency_table = {}
    number_of_apps = len(dataset)
    for app in dataset:
        genre = app[index]
        if genre not in frequency_table:
            frequency_table[genre] = 1
        else:
            frequency_table[genre] += 1
    for key in frequency_table:
        frequency_table[key] = (frequency_table[key]/number_of_apps)*100
    return frequency_table

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Lets strart with AppStore dataset:

In [55]:
display_table(ios_free_eng, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The most popular genre is `games` with 58% from total. The next is `entertainment` with just 7%. `entertainment` apps and `games` are of the same nature, people use those for fun and probably not in practical purposes. In my oppinion amount of apps in genre is not so reliable marker for our end goal. But for a moment we will kepp it as it is in guide for project.

In [56]:
display_table(android_final, 1)

FAMILY : 18.774681259167323
GAME : 9.635563578923614
TOOLS : 8.450863138892023
BUSINESS : 4.580841701455489
LIFESTYLE : 3.9038700214374367
PRODUCTIVITY : 3.8925871601038025
FINANCE : 3.7007785174320205
MEDICAL : 3.542818458761142
SPORTS : 3.4187069840911652
PERSONALIZATION : 3.317161232088458
COMMUNICATION : 3.2494640640866526
HEALTH_AND_FITNESS : 3.068938282748505
PHOTOGRAPHY : 2.944826808078529
NEWS_AND_MAGAZINES : 2.798149610741284
SOCIAL : 2.6627552747376737
TRAVEL_AND_LOCAL : 2.335552296062281
SHOPPING : 2.245289405393208
BOOKS_AND_REFERENCE : 2.1437436533904997
DATING : 1.8616721200496444
VIDEO_PLAYERS : 1.7826920907142052
MAPS_AND_NAVIGATION : 1.399074805370642
EDUCATION : 1.2523976080333974
FOOD_AND_DRINK : 1.241114746699763
ENTERTAINMENT : 1.0380232426943474
LIBRARIES_AND_DEMO : 0.9364774906916393
AUTO_AND_VEHICLES : 0.9251946293580051
HOUSE_AND_HOME : 0.8349317386889316
WEATHER : 0.8010831546880289
EVENTS : 0.7108202640189552
ART_AND_DESIGN : 0.6769716800180525
PARENTING : 0.

In GooglePlay store dataset apps are more evenly spreaded in `category`. But numer of those categories is much bigger. Also remarkable that pattern is different from the one in AppStore dataset. Categories `tools` and `busines` have approximately the same percentage as `games`. Hipotesis that genres and categories is bad foundation for judjement is getting more reasonable.

Lets move to last column to analise in this part of research.

In [57]:
display_table(android_final, 9)

Tools : 8.43958027755839
Entertainment : 6.070179397495204
Education : 5.348076272142616
Business : 4.580841701455489
Productivity : 3.8925871601038025
Lifestyle : 3.8925871601038025
Finance : 3.7007785174320205
Medical : 3.542818458761142
Sports : 3.463838429425702
Personalization : 3.317161232088458
Communication : 3.2494640640866526
Action : 3.102786866749408
Health & Fitness : 3.068938282748505
Photography : 2.944826808078529
News & Magazines : 2.798149610741284
Social : 2.6627552747376737
Travel & Local : 2.324269434728647
Shopping : 2.245289405393208
Books & Reference : 2.1437436533904997
Simulation : 2.042197901387792
Dating : 1.8616721200496444
Arcade : 1.8616721200496444
Video Players & Editors : 1.7826920907142052
Casual : 1.7488435067133026
Maps & Navigation : 1.399074805370642
Food & Drink : 1.241114746699763
Puzzle : 1.128286133363421
Racing : 0.9928917973598104
Role Playing : 0.9364774906916393
Libraries & Demo : 0.9364774906916393
Auto & Vehicles : 0.9251946293580051
Str

The most frequently encountered genre in GooglePlay dataset is `Tools`. It is absolutly opposit genre to one we saw in AppStore. As for now it seems that IOs developers more focused on gaming apps, while android developers more about usefull apps. But again, as I see, it is impracticall approach to analise genres to find out which app is likely to attract more users. This might be intermediate step to move further and find out how many users each genre attracts.

Project's guide adviced a bit different idea. Basicaly it is the same, but insted of total amount of users it suggested to count average amount of users per app in genre. Probably this is more representative approach. It provides clear view on how app genre affects amount of users.

We can easily obtain required data (amount of users) for GooglePlay data set, it is contained in `installs` column. For AppStore dataset we can use `rating_count_tot` as proxy to find out amount of users. It is not accurate figure, but as we use it in scope of one dataset to compare data within AppStore - we can presume that percentage of users tend to leave app rating is the same for every app in data set. This approach will anihilate error within dataset, but we can not compare figures between IOs and android apps. Comparison posible only between parts and percentages.

Let's start with GooglePlay data set.

In [90]:
def average_users_per_genre(dataset, index_genre, index_users):
    frequency_table = {}
    for app in dataset:
        genre = app[index_genre]
        if genre not in frequency_table:
            frequency_table[genre] = 1
        else:
            frequency_table[genre] += 1   
    users_per_genre = {}
    for key in frequency_table:
        for app in dataset:
            users_num = app[index_users]
            users_num = users_num.replace(',', '')
            users_num = users_num.replace('+', '')
            if app[index_genre] == key and app[index_genre] not in users_per_genre:
                users_per_genre[key] = float(users_num)
            elif app[index_genre] == key and app[index_genre] in users_per_genre:
                users_per_genre[key] += float(users_num)
    average_users_per_genre = []
    for key in frequency_table:
        average_users_per_genre.append([round(float(users_per_genre[key]/frequency_table[key])), key])
    sorted_average_users_per_genre = sorted(average_users_per_genre, reverse = True)
    for pair in sorted_average_users_per_genre:
        print (pair[1], ':', pair[0])

Code above will represent genres with correspondig amount of users. Let's start with AppStore dataset:

In [91]:
average_users_per_genre(ios_free_eng, 11, 5)

Navigation : 86090
Reference : 74942
Social Networking : 71548
Music : 57327
Weather : 52280
Book : 39758
Food & Drink : 33334
Finance : 31468
Photo & Video : 28442
Travel : 28244
Shopping : 26920
Health & Fitness : 23298
Sports : 23009
Games : 22789
News : 21248
Productivity : 21028
Utilities : 18684
Lifestyle : 16486
Entertainment : 14030
Business : 7491
Education : 7004
Catalogs : 4004
Medical : 612


First assesment shows that apps with largest amount of apps are the onec easy to develop and probably market already filed with other mighty companies. First five is: Navigation, Reference, Social Networking, Music, Weather. For our company the success recepie is to find genre with large user amount, market not floded with similar apps, and there is no need for other infrostructure for app to work (we are not going to employ large amount of finantial analitics to make Finance app). I would recomend to develop `Photo&Video`, `Food&Drinks`, `Book` apps for AppStore market. Those apps are in top 10 averege users list and there are not so much of those on market. The only exception is Photo and Video, those apps are popular and there is no lack on market, but those seem to be on boarder with Social Network apps. If our developers will find way to integrate apps with social networks it will widely expand amount of users. 

Lets check android data set:

In [93]:
average_users_per_genre(android_final, 1, 5)

COMMUNICATION : 38326063
VIDEO_PLAYERS : 24790074
SOCIAL : 23253652
ENTERTAINMENT : 19428913
PHOTOGRAPHY : 17805628
PRODUCTIVITY : 16772839
TRAVEL_AND_LOCAL : 13984078
GAME : 13006873
TOOLS : 10814477
NEWS_AND_MAGAZINES : 9549178
BOOKS_AND_REFERENCE : 8767812
SHOPPING : 7036877
PERSONALIZATION : 5201483
WEATHER : 5074486
FAMILY : 4374336
SPORTS : 4274689
HEALTH_AND_FITNESS : 4167457
MAPS_AND_NAVIGATION : 4056942
EDUCATION : 3057207
FOOD_AND_DRINK : 1924898
ART_AND_DESIGN : 1905352
BUSINESS : 1704192
LIFESTYLE : 1437816
FINANCE : 1387692
HOUSE_AND_HOME : 1313682
DATING : 854029
COMICS : 817657
AUTO_AND_VEHICLES : 647318
LIBRARIES_AND_DEMO : 638504
PARENTING : 542604
BEAUTY : 513152
EVENTS : 253542
MEDICAL : 107144


As we aimed to find apps which will be profitable for both markets I will check app genres which I have suggested for AppStore. 
* `Photo&Video` - there is category `photography` in android dataset. It might be interesting for our team, cause at the moment huge amount of content generated and if we are able to transfer user from real world to imagery using masks or backgrounds it could be profitable. It is in top 5 of list `average amount of users`, but we will get quite some chalandge with other apps.
* `Food&Drinks` - this app genre is not on top in `average amount of users`, but on other hand there is not so much chalanges with other apps.
* `Books` - it is on 11th place in list of `average amount of users`, and there is not so much other apps in this genre. It is golden meedle for us and the best option to develop. It could be app which allowes you to read reviews, or app with quest on popular book, or any other fun-app based on popular book.